lib/internal/errors.js

```diff
@@ -40,9 +40,42 @@ const prepareStackTrace = (globalThis, error, trace) => {
   if (trace.length === 0) {
     return errorString;
   }
-  return `${errorString}\n    at ${trace.join('\n    at ')}`;
+
+  const { getOptionValue } = require('internal/options');
+  if (getOptionValue('--experimental-source-maps')) {
+    return `${errorString}\n    at ${annotateWithSourceMaps(trace)}`;
+  } else {
+    return `${errorString}\n    at ${trace.join('\n    at ')}`;
+  }
 };
 
+function annotateWithSourceMaps(trace) {
+  const { fetchSourceMap } = require('internal/source_map/source_map_cache');
+  const { SourceMap } = require('internal/source_map/source_map');
+  const debug = require('internal/util/debuglog').debuglog('source_map');
+  let annotated = '';
+  try {
+    for (let i = 0, t; (t = trace[i]) !== undefined; i++) {
+      if (i !== 0) annotated += '\n    at ';
+      annotated = `${annotated} ${t}`;
+
+      const sourceMap = fetchSourceMap(t.getFileName());
+      if (sourceMap && sourceMap.payload) {
+        const sm = new SourceMap(sourceMap.payload);
+        const [, , sourceURL, sourceLineNumber, sourceColumnNumber] =
+                           sm.findEntry(t.getLineNumber(), t.getColumnNumber());
+        if (sourceURL && sourceLineNumber !== undefined &&
+            sourceColumnNumber !== undefined) {
+          annotated +=
+`\n         -> ${sourceURL.replace('file://', '')}:${sourceLineNumber}:${sourceColumnNumber}`;
```

Comment:
I love the idea too, it was actually @hashseed who brought up the idea that we keep both in the stack-trace; it feels like the best of both worlds.

Reply:
This will probably break tools that do long stack trace stitching (like bluebird for example) but that's a bridge we can cross later and I also like this idea.